src/_pytest/main.py

```diff
@@ -958,16 +964,41 @@ def collect(self) -> Iterator[nodes.Item | nodes.Collector]:
             self.trace.root.indent -= 1
 
     def genitems(self, node: nodes.Item | nodes.Collector) -> Iterator[nodes.Item]:
+        import inspect
+
+        from _pytest.python import Class
+        from _pytest.python import Function
+        from _pytest.python import Module
+
         self.trace("genitems", node)
         if isinstance(node, nodes.Item):
             node.ihook.pytest_itemcollected(item=node)
+            if not self.config.getini("collect_imported_tests"):
```

Comment:
Previously this code was placed in `_pytest.python`, which I think is the correct place... as it is implemented here, we are calling `pytest_itemcollected` but then discarding the item if the option is set and the item is not defined in its module, which I think is incorrect: `pytest_itemcollected` should only really be called if we did collect the item.

In fact, it is important for this to be tested too: the following collection hooks should not receive any item that was discarded due to `collect_imported_tests`:

* `pytest_itemcollected`
* `pytest_collectreport`
* `pytest_collection_modifyitems`

Reply:
Hi! Thought I would pick up this issue again after some wait. Is there any examples on how I could test these hooks? Im unsure on how to extract that information from the hooks. Im assuming this is problematic because extensions to pytest could be listening to these hooks?

Reply:
> Is there any examples on how I could test these hooks? Im unsure on how to extract that information from the hooks.

Usually we write a `conftest.py` file in the tests that implement these hooks, and then set/update some global variable when they are called.

> Im assuming this is problematic because extensions to pytest could be listening to these hooks?

Exactly.